packages/widget.client-react/src/client.test.tsx

```diff
@@ -0,0 +1,269 @@
+      },
+      resolveAtIndex(index: number) {
+        const item = pending[index];
+        if (item) {
+          pending.splice(index, 1);
+          item.resolve(item.payload);
+        }
+      },
+    };
+  };
+
+  beforeEach(() => {
+    vi.clearAllMocks();
+    vi.useFakeTimers();
+    mockWidgetClient.hostEventTarget = new FoundryHostEventTarget();
+  });
+
+  afterEach(() => {
+    unmountWidget();
+    vi.useRealTimers();
+  });
+
+  it("should emit both events when different events are called", async () => {
+    const deferred = createDeferredTransformMock();
+    const emit = await mountWidget();
+
+    act(() => {
+      emit("eventA", { parameterUpdates: { myParam: "valueA" } });
+      emit("eventB", { parameterUpdates: { myParam: "valueB" } });
+    });
+
+    expect(deferred.pendingCount).toBe(2);
+
+    await act(async () => {
+      deferred.resolveAll();
+      await vi.runAllTimersAsync();
+    });
+
+    // Both events should be emitted
+    expect(mockEmitEvent).toHaveBeenCalledTimes(2);
+    expect(mockEmitEvent).toHaveBeenCalledWith("eventA", {
+      parameterUpdates: { myParam: "valueA" },
+    });
+    expect(mockEmitEvent).toHaveBeenCalledWith("eventB", {
+      parameterUpdates: { myParam: "valueB" },
+    });
+  });
+
+  it("should only emit the last call even when transforms resolve out of order", async () => {
+    const deferred = createDeferredTransformMock();
+    const emit = await mountWidget();
+
+    // Call emitEvent 3 times rapidly
+    act(() => {
+      emit("eventA", { parameterUpdates: { myParam: "first" } });
+      emit("eventA", { parameterUpdates: { myParam: "second" } });
+      emit("eventA", { parameterUpdates: { myParam: "third" } });
+    });
+
+    expect(deferred.pendingCount).toBe(3);
+
+    // Resolve in reverse order: third, then first, then second
+    await act(async () => {
+      deferred.resolveAtIndex(2); // "third" resolves first
+      await vi.runAllTimersAsync();
+    });
+
+    // The third call should emit immediately since it's the latest
+    expect(mockEmitEvent).toHaveBeenCalledTimes(1);
+    expect(mockEmitEvent).toHaveBeenCalledWith("eventA", {
+      parameterUpdates: { myParam: "third" },
+    });
+
+    await act(async () => {
+      deferred.resolveAtIndex(0); // "first" resolves second
+      deferred.resolveAtIndex(0); // "second" resolves last (now at index 0)
+      await vi.runAllTimersAsync();
+    });
+
+    // Still only one emit - the stale calls are ignored
+    expect(mockEmitEvent).toHaveBeenCalledTimes(1);
+  });
+
+  it("should emit all calls when they complete sequentially", async () => {
+    const deferred = createDeferredTransformMock();
+    const emit = await mountWidget();
+
+    // First call
+    act(() => {
+      emit("eventA", { parameterUpdates: { myParam: "first" } });
+    });
+
+    // Complete first call before making second
+    await act(async () => {
+      deferred.resolveNext();
+      await vi.runAllTimersAsync();
+    });
```

Comment:
From the other comment is this actually representative of the real world? It seems hopeful but not guaranteed that the promise will complete before another event could be emitted. I think we'd definitely want a predictable result if "first" and "second" are set in order to not get different results each time